File: Python/Adjust_Mean_Sample.py
import numpy as np

# Given data
values = np.array([2, 4, 10, 20, 50, 4, 5, 6, 9, 80], dtype=float)
current_avg = np.mean(values)
target_avg = 25  # Try with a value below or above current_avg
num_to_adjust = max(1, int(round(0.2 * len(values))))  # At least 1 value
max_adjustment = 100  # Maximum allowed adjustment per value

# Calculate total sum difference needed
total_sum_needed = target_avg * len(values)
sum_difference = total_sum_needed - np.sum(values)

# Decide which values to adjust
if sum_difference > 0:
    # Need to increase: adjust the lowest values
    adjust_indices = np.argsort(values)[:num_to_adjust]
else:
    # Need to decrease: adjust the highest values
    adjust_indices = np.argsort(values)[-num_to_adjust:]

# Track how much each value has been adjusted
adjusted_amounts = {idx: 0.0 for idx in adjust_indices}

remaining_diff = sum_difference
remaining_indices = list(adjust_indices)

while remaining_indices and abs(remaining_diff) > 1e-8:
    per_value = remaining_diff / len(remaining_indices)
    to_remove = []
    for idx in remaining_indices:
        # Calculate how much more this value can be adjusted (positive or negative)
        remaining_adj = max_adjustment - abs(adjusted_amounts[idx])
        # Cap the adjustment to max_adjustment per value
        adj = np.clip(per_value, -remaining_adj, remaining_adj)
        # Ensure no negative values
        if values[idx] + adj < 0:
            adj = -values[idx]
        # Final check: do not exceed max_adjustment in total
        if abs(adjusted_amounts[idx] + adj) > max_adjustment:
            adj = np.sign(adj) * (max_adjustment - abs(adjusted_amounts[idx]))
        values[idx] += adj
        adjusted_amounts[idx] += adj
        remaining_diff -= adj
        # Remove if max_adjustment reached or value is zero
        if abs(adjusted_amounts[idx]) >= max_adjustment or values[idx] == 0:
            to_remove.append(idx)
    for idx in to_remove:
        remaining_indices.remove(idx)
    if not to_remove:
        break  # No more adjustments possible

new_avg = np.mean(values)

# Output results
print("Adjusted Values:", values)
print("Total adjustments per index:", adjusted_amounts)
print("New Average:", new_avg)
print("Max adjustment actually used:", max(abs(a) for a in adjusted_amounts.values()))

File: C++/Adjust_Mean_Sample.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <cmath>

int main() {
    // Given data
    std::vector<double> values = {2, 4, 10, 20, 50, 4, 5, 6, 9, 80};
    double current_sum = std::accumulate(values.begin(), values.end(), 0.0);
    double current_avg = current_sum / values.size();
    double target_avg = 15; // Try with a value below or above current_avg
    int num_to_adjust = std::max(1, static_cast<int>(std::round(0.2 * values.size())));

    // Calculate total sum difference needed
    double total_sum_needed = target_avg * values.size();
    double sum_difference = total_sum_needed - current_sum;

    // Decide which values to adjust
    std::vector<int> adjust_indices(values.size());
    std::iota(adjust_indices.begin(), adjust_indices.end(), 0);

    if (sum_difference > 0) {
        // Need to increase: adjust the lowest values
        std::sort(adjust_indices.begin(), adjust_indices.end(),
            [&values](int i, int j) { return values[i] < values[j]; });
    } else {
        // Need to decrease: adjust the highest values
        std::sort(adjust_indices.begin(), adjust_indices.end(),
            [&values](int i, int j) { return values[i] > values[j]; });
    }

    adjust_indices.resize(num_to_adjust);

    // Distribute the difference among chosen values, ensuring no negative values
    double remaining_diff = sum_difference;
    std::vector<int> remaining_indices = adjust_indices;

    while (!remaining_indices.empty() && std::abs(remaining_diff) > 1e-8) {
        double per_value = remaining_diff / remaining_indices.size();
        std::vector<int> to_remove;
        for (int idx : remaining_indices) {
            if (values[idx] + per_value < 0) {
                // Only bring it down to zero
                remaining_diff -= values[idx];
                values[idx] = 0;
                to_remove.push_back(idx);
            } else {
                values[idx] += per_value;
                remaining_diff -= per_value;
            }
        }
        // Remove indices that have been set to zero
        for (int idx : to_remove) {
            remaining_indices.erase(std::remove(remaining_indices.begin(), remaining_indices.end(), idx), remaining_indices.end());
        }
        if (to_remove.empty()) break; // No more negatives, done
    }

    // Calculate new average
    double new_avg = std::accumulate(values.begin(), values.end(), 0.0) / values.size();

    // Output results
    std::cout << "Adjusted Values: ";
    for (double v : values) std::cout << v << " ";
    std::cout << std::endl;
    std::cout << "New Average: " << new_avg << std::endl;

    return 0;
}